### PR TITLE
[SPARK-2429] [MLlib] Hierarchical Implementation of KMeans

### PR DESCRIPTION
**Pull request by yu-iskw**
I want to add a divisive hierarchical clustering algorithm implementation to MLlib. I don't support distance metrics other Euclidean distance metric yet. It would be nice to support it at other issue.
Could you review it?

Thanks!

### CODE REVIEW

**Comment by rnowling** [comment]
@mengxr This PR has been lingering for a while.  What can we do to get it a little more attention?  Thanks!

**Comment by mengxr** [comment]
@yu-iskw @rnowling, I asked @freeman-lab to make one pass on this PR. Let's ping him :)

**Comment by freeman-lab** [comment]
Hey all, thanks for the nudge =) I've been going through it, will get you feedback ASAP.

**Comment by rnowling** [comment]
Thanks @mengxr @freeman-lab! :)

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala [hunk 1]
```diff
@@ -0,0 +1,627 @@
...
+object ClusterTree {
+
+  /**
+   * Converts `RDD[Vector]` into a ClusterTree instance
+   *
+   * @param data the data in a cluster
+   * @return a ClusterTree instance
+   */
+  def fromRDD(data: RDD[Vector]): ClusterTree = {
+    val breezeData = data.map(_.toBreeze).cache
```
Comment (freeman-lab) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala]: This automatically caches the user's input data. Elsewhere, like in `KMeans`, the behavior is to issue a warning if the input data is not cached, but not to automatically cache it for them. Because you are also automatically caching (and then unpersisting) all split data sets throughout the algorithm, this might be ok, but there should be a note in the doc string explaining that data will be cached by this algorithm.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+    }
+
+    val vectors = centers.map(center => Vectors.fromBreeze(center))
+    val nodes = centers.size match {
+      case 1 => Array(new ClusterTree(vectors(0), data))
+      case 2 => {
+        val closest = data.map(p => (ClusterTree.findClosestCenter(metric)(centers)(p), p))
+        centers.zipWithIndex.map { case (center, i) =>
+          val subData = closest.filter(_._1 == i).map(_._2)
+          subData.cache
```
Comment (freeman-lab) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala]: Is it correct that caching every split (but then unpersisting) like this can potentially end up caching twice the size of the original data set? If so, that should be made very clear in the documentation, as it may affect the user's choice of what size data sets to run this one (and their hardware or instance provisioning).

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+
+  def setRandomRange(range: Double): this.type = {
+    this.randomRange = range
+    this
+  }
+}
+
+
+/**
+ * This is a divisive hierarchical clustering algorithm based on bi-sect k-means algorithm.
```
Comment (freeman-lab) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala]: Give a slightly longer overview of how the algorithm works?

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClusteringModel.scala [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+   * @return List[(node1, node2, distance, tree size)]
+   */
+  def toMergeList(): List[(Int, Int, Double, Int)] = {
+    val seq = this.clusterTree.toSeq().sortWith{ case (a, b) => a.getHeight() < b.getHeight()}
+    val leaves = seq.filter(_.isLeaf())
+    val nodes = seq.filter(!_.isLeaf()).filter(_.children.size > 1)
+    val clusters = leaves ++ nodes
+    val treeMap = clusters.zipWithIndex.map { case (tree, idx) => (tree -> idx)}.toMap
+
+    // If a node only has one-child, the child is regarded as the cluster of the child.
```
Comment (freeman-lab) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClusteringModel.scala]: I find this description a little hard to follow.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClusteringModel.scala [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+   */
+  def getSumOfVariance(): Double = this.getClusters().map(_.getVariance().get).sum
+
+  def getClusters(): Array[ClusterTree] = clusterTree.getClusters().toArray
+
+  def getCenters(): Array[Vector] = getClusters().map(_.center)
+
+  /**
+   * Converts a clustering merging list
+   * Returned data format is fit for scipy's dendrogram function
```
Comment (freeman-lab) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClusteringModel.scala]: I think it's a little weird to justify this based on a connection to scipy, and to reference that code so explicitly. This is primarily scala code, after all =) More importantly, the basic logic of this data structure is quite general, and is used in at least scipy and matlab (and possibly also R?). I'd instead give a longer description of how the list is organized here in the doc, and maybe mention that it is used by other libraries.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClusteringModel.scala [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+   * SEE ALSO: scipy.cluster.hierarchy.dendrogram
+   *
+   * @return List[(node1, node2, distance, tree size)]
+   */
+  def toMergeList(): List[(Int, Int, Double, Int)] = {
```
Comment (freeman-lab) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClusteringModel.scala]: Consider renaming -> `toLinkageMatrix`? I think that's a more general term for this data structure. Would require renaming here and elsewhere (e.g. in the Python code).

File: python/pyspark/mllib/clustering.py [hunk 7]
```diff
@@ -88,6 +92,162 @@ def train(cls, rdd, k, maxIterations=100, runs=1, initializationMode="k-means||"
...
+        sc = rdd.context
+        ser = PickleSerializer()
+        # cache serialized data to avoid objects over head in JVM
+        cached = rdd.map(_convert_to_vector)._reserialize(AutoBatchedSerializer(ser)).cache()
+        model = sc._jvm.PythonMLLibAPI().trainHierarchicalClusteringModel(
+            _to_java_object_rdd(cached), k,
+            subIterations, numRetries, epsilon, randomSeed, randomRange)
+        bytes = sc._jvm.SerDe.dumps(model.getCenters())
+        centers = ser.loads(str(bytes))
+        # TODO: because centers are SparseVector, wel will change them numpy.array.
```
Comment (freeman-lab) [inline python/pyspark/mllib/clustering.py]: Did you mean convert from `DenseVector`? That's currently the output. I agree it would be good to change to `numpy.array`, especially because after calling `cut` the output is an array of `numpy.array`s. The two behaviors should be consistent, as calling `cut` should not change the type.

File: docs/mllib-clustering.md [hunk 8]
```diff
@@ -154,6 +156,175 @@ section of the Spark
...
+
+### Hierarchical Clustering Example
+
+<div class="codetabs">
+
+<div data-lang="scala" markdown="1">
+The following code snippets can be executed in `spark-shell`.
+
+In the following example after loading and parsing data, 
+we use the hierarchical clustering object to cluster the sample data into three clusters.
```
Comment (freeman-lab) [inline docs/mllib-clustering.md]: Clarify that this means three clusters at the bottom-most levels of a hierarchical tree.

File: examples/src/main/python/mllib/hierarchical_clustering.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+A hierarchical clustering program using MLlib.
+
+This example requires NumPy, SciPy and matplotlib.
+"""
+
+import os
+import sys
+
+from numpy import array
+import matplotlib.pyplot as plt
```
Comment (freeman-lab) [inline examples/src/main/python/mllib/hierarchical_clustering.py]: I love that you've made it so easy to visualize the output, but this now adds a "dependency" on matplotlib which isn't used anywhere else in PySpark AFAIK. Strictly, because PySpark doesn't currently use formal package management (e.g. through PyPi), this isn't really adding a dependency, and it's just an example. But might be safer to just use a line note showing how the output can be visualized with matplotlib. Curious what others think. cc @davies

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala [hunk 10]
```diff
@@ -0,0 +1,627 @@
...
+}
+
+
+/**
+ * A cluster as a tree node which can have its sub nodes
+ *
+ * @param center the center of the cluster
+ * @param data the data in the cluster
+ * @param height distance between sub nodes
+ * @param variance the statistics for splitting of the cluster
```
Comment (freeman-lab) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala]: "of the cluster" => "the cluster"

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala [hunk 11]
```diff
@@ -0,0 +1,627 @@
...
+    }
+    seq.sortWith { case (a, b) =>
+      a.getDepth() < b.getDepth() &&
+          breezeNorm(a.center.toBreeze, 2) < breezeNorm(b.center.toBreeze, 2)
+    }
+  }
+
+  /**
+   * Gets the all clusters which are leaves in the cluster tree
+   * @return the Seq of the clusters
```
Comment (freeman-lab) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala]: Insert line break

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala [hunk 12]
```diff
@@ -0,0 +1,627 @@
...
+  def getHeight(): Double = {
+    this.children.size match {
+      case 0 => 0.0
+      case _ => this.height.get + this.children.map(_.getHeight()).max
+    }
+  }
+
+  /**
+   * Assigns the closest cluster with a vector
+   * @param metric distance metric
```
Comment (freeman-lab) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala]: Insert line break

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala [hunk 13]
```diff
@@ -0,0 +1,627 @@
...
+        val distances = this.children.map(tree => metric(tree.center.toBreeze, v.toBreeze))
+        val minIndex = distances.indexOf(distances.min)
+        this.children(minIndex).assignCluster(metric)(v)
+      }
+    }
+  }
+
+  /**
+   * Assigns the closest cluster index of the clusters with a vector
+   * @param metric distance metric
```
Comment (freeman-lab) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/HierarchicalClustering.scala]: Insert line break

**Comment by freeman-lab** [comment]
Hi @yu-iskw and @rnowling , I've spent time reviewing the code and using it in both Python and Scala. Overall great work, terrific to see my little gist turned into something so refined and performant! =) I left lots of comments, most minor, though documenting the caching behavior seems quite important.

The one significant addition I'd suggest is exposing another model output: a list of the centers at all nodes in the learned tree. This would be in addition to just the centers of the leaves, which is currently returned by `getCenters` (or `clusterCenters` in Python). Maybe call it `getTreeCenters`. It's basically given by `model.clusterTree.toSeq().map(_.center)`. But we should make sure it's sorted so that it can be indexed using the values from the merge list. In other words, if `Z` is the merge list, and row i indicates that `Z[i,0]` and `Z[i,1]` were merged, we want to be able to get the centers associated with those nodes by calling, for example, `model.treeCenters[Z[i,0]]` and `model.treeCenters[Z[i,1]]`. What do you think?

File: examples/src/main/python/mllib/hierarchical_clustering.py (hunk 9 above)
Comment (davies) [inline examples/src/main/python/mllib/hierarchical_clustering.py]: We should be careful to add any dependency, even in example. For here, I'd like to make it optional, tell user to install it to get a better experience if matplotlib and scipy is not installed.

File: data/mllib/sample_hierarchical_data.csv [hunk 14]
```diff
@@ -0,0 +1,150 @@
+5.1,3.5,1.4,0.2
```
Comment (srowen) [inline data/mllib/sample_hierarchical_data.csv]: Minor point - this wouldn't really be CSV though. I imagine the example shows parsing a common encoding like this on purpose.

Comment (freeman-lab) [inline data/mllib/sample_hierarchical_data.csv]: Good point =) Leave as is then. Maybe at some point we should give all the vector-valued example data sets the same format / file type just for consistency, but that can be a separate PR.

File: docs/mllib-clustering.md [hunk 15]
```diff
@@ -154,6 +156,175 @@ section of the Spark
...
+import org.apache.spark.SparkConf;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.api.java.function.Function;
+import org.apache.spark.mllib.clustering.HierarchicalClustering;
+import org.apache.spark.mllib.clustering.HierarchicalClusteringModel;
+import org.apache.spark.mllib.linalg.Vector;
+import org.apache.spark.mllib.linalg.Vectors;
+
+public class JavaHierarchicalClustering {
```
Comment (srowen) [inline docs/mllib-clustering.md]: The other example code I see foregoes a lot of the boilerplate here of declaring a class, main method, System.out, etc. The indentation here is also significantly deeper than the 2-space indent in the code. Addressing these might make it easier to scan as an example on the web page.

File: examples/src/main/java/org/apache/spark/examples/mllib/JavaHierarchicalClustering.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.examples.mllib;
+
+import org.apache.spark.SparkConf;
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/mllib/JavaHierarchicalClustering.java]: These look correctly ordered in the sense that package `a.b.c` sorts entirely before `a.b.c.d`

File: mllib/src/test/java/org/apache/spark/mllib/clustering/JavaHierarchicalClusteringSuite.java [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.Serializable;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+public class JavaHierarchicalClusteringSuite implements Serializable {
+  private transient JavaSparkContext sc;
```
Comment (srowen) [inline mllib/src/test/java/org/apache/spark/mllib/clustering/JavaHierarchicalClusteringSuite.java]: This is not a comment on this PR per se, but this whole `implements Serializable` and `transient JavaSparkContext` thing is an anti-pattern I wish wasn't used in even the tests.

File: mllib/src/test/java/org/apache/spark/mllib/clustering/JavaHierarchicalClusteringSuite.java [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+        Vectors.dense(1.0, 2.0, 6.0),
+        Vectors.dense(1.0, 3.0, 0.0),
+        Vectors.dense(1.0, 4.0, 6.0)
+    );
+    JavaRDD<Vector> data = sc.parallelize(points, 2);
+    HierarchicalClustering algo = new HierarchicalClustering().setNumClusters(1);
+    HierarchicalClusteringModel model = algo.run(data.rdd());
+    JavaRDD<Integer> predictions = model.predict(data);
+    // Should be able to get the first prediction.
+    predictions.first();
```
Comment (srowen) [inline mllib/src/test/java/org/apache/spark/mllib/clustering/JavaHierarchicalClusteringSuite.java]: assert what the first one is? or is it not stable enough to reliably test for?

**Comment by rnowling** [comment]
@freeman-lab @srowen @mengxr many thanks!

**Comment by yu-iskw** [comment]
@freeman-lab, @srowen, I apologize for the delay in replying. I will modify the code ASAP.
And I have a question about the implementation. I think this implementation is very slow and it difficult to take the large number of clusters in an argument. So, I tried to implement the new one which is more scalable and faster than current one. The new one is 1000 times faster than the current one.

https://github.com/yu-iskw/more-scalable-hierarchical-clustering-with-spark

Should we continue the PR, or replace the current one with the new one. thanks!

**Comment by yu-iskw** [comment]
I've spoken with @freeman-lab. I am going to send a new PR after replacing the algorithm to the new one and adding wrapper classes for ml package.
